### ISSUE TITLE
Question about using autorun

### ISSUE THREAD

**Issue by dnish** [mobxjs/mobx#1010]
Hey,
I have a little question about using autorun with different patterns (in my case with and without async/await). 

For example, I'm having the following 2 variants:

     autorun(() => {

            if (this.currentUploads.length < this.limit && this.uploadsQueue.length > 0) {
                const uploadFile = this.uploadsQueue[0];
                this.startUpload(uploadFile);
            }

     });

    startUpload(uploadFile) {

        this.uploadsQueue = _.filter(this.uploadsQueue, (file) => file._id !== uploadFile._id); // Need to use _.find because _.remove mutates the observable array which causes errors.
        if (_.find(this.currentUploads, (file) => file._id === uploadFile._id)) return; // Prevent duplicates
        this.currentUploads.push(uploadFile);

        RNFetchBlob.fs.exists(uploadFile.path).then((fileExists) => { .....

vs

     autorun(() => {

            if (this.currentUploads.length < this.limit && this.uploadsQueue.length > 0) {
                const uploadFile = this.uploadsQueue[0];
                this.startUpload(uploadFile);
            }

     });

    async startUpload(uploadFile) {

        this.uploadsQueue = _.filter(this.uploadsQueue, (file) => file._id !== uploadFile._id); // Need to use _.find because _.remove mutates the observable array which causes errors.
        if (_.find(this.currentUploads, (file) => file._id === uploadFile._id)) return; // Prevent duplicates
        this.currentUploads.push(uploadFile);

        const fileExists = await RNFetchBlob.fs.exists(uploadFile.path);
        ...

Is this the correct execution order?

Variant 1:

1. New value for this.currentUploads
2. Autorun reruns
3. Autorun waits until the last synchronous action of startUpload has finished. In my case it would be <code>this.currentUploads.push(uploadFile);</code>. After that, we get a Promise (<code>RNFetchBlob.fs.exists</code>).
4. Rerun if another value changes

Variant 2:

1. New value for this.currentUploads
2. Autorun reruns
3. Rerun if another value changes

On variant 2 <code>startUpload</code> is a Promise, so it would not block. The functions finishes instantly, without waiting for the duplicate check.

I'm just wondering about the behavior if we get multiple changes within a small time window. Or is autorun asynchronous, so variant 1 wouldn't block and multiple reruns could happen at the same time (without waiting that the other has finished)?

**Comment by urugator** [comment]
```javascript 
// ouside of action
this.state = "a"; // immediately (synchronously) invokes reactions/autoruns

// inside the action
runInAction(() => {
  this.state1 = "a"; // nothing happens
  this.state1 = "b"; // nothing happens  
  runInAction(() => {
    this.state2 = "a"; // nothing happens
    this.state2 = "b"; // nothing happens
    // end of the nested action
    // nothing happens    
  })
  // end of the outermost action
  // the reactions/autoruns are immediately synchronously invoked with the current state
  // the number of modifications inside action doesn't matter       
})
```
So if you modify observable inside an autorun (which you should not) all autoruns(including this one) and reactions will be invoked synchronously at the end of the current outermost `action` or immediately if there isn't any (possibly causing an endless loop).

Any asynchronous code executed from autorun technically doesn't run inside the autorun itself, so any observables accessed are not tracked by autorun.

Does it answer the question?

**Comment by urugator** [comment]
Bonus for inspiration:
```javascript  
class BufferedString {
  constructor() {
    this.value = "";
    this.buffer = "";
    this.threshold = 3;
    this.atom = new Mobx.Atom("BufferedString");  
  }
  append(value) {    
    this.buffer += value;
    if (this.buffer.length > this.threshold) {
      this.value += this.buffer; // update value
      this.buffer = ""; // clear buffer
      this.atom.reportChanged(); // notify observers
    }    
  } 
  get() {
    this.atom.reportObserved(); // notify MobX about usage
    return this.value;
  }
}

const bs = new BufferedString();
Mobx.autorun(() => console.log(bs.get());
[..."123456789"].forEach(char => bs.append(char))

/*
Output:
123
123456
123456789
*/
```

**Comment by dnish** [comment]
@urugator Thank you very much :)

    So if you modify observable inside an autorun (which you should not)

Oh okay, is there any reason for it (except that it could cause an infinite loop)? In our case, we scan 2 observables and modify another if one of them has changed. We use this pattern to enable a upload queue and concurrent upload limit (in our case 3 active uploads). Just thought that autorun is perfect for this, because it detects when a upload has finished and can start the next one from the queue.

**Comment by urugator** [comment]
If the modifications (shifting the queue or whatever) is done asynchronously after the upload is finishes, then it's fine, because it doesn't actually happen inside the autorun itself and doesn't affect the current autorun execution.
But synchronous recursive behavior implemented with observables (or events or any middleware) is hard to follow, because the recursion is not immediately apparent. You don't see when, what and why is causing the nested call and blocking the current one. Additionaly you have to deal with disposers or subscription/unsubscriptions and other mental burden, which is absolutely unnecessary unless you need to decouple some system parts.
There is a reason why you can't modify state in React's render method and create a possibly inifinite recursion ... it's simply hard to reason about such system.

**Comment by mweststrate** [comment]
@dnish if you intend to modify observables in autorun, I would at least use `reaction` instead of `autorun`, as it makes more explicit what your code is reacting to, and what the effect will be.
